options/posix/generic/services.cpp

```diff
@@ -11,7 +11,88 @@
 
 namespace mlibc {
 
-static int lookup_serv_file(struct service_buf *buf, const char *name,
+static int parse_rest(service_buf &buf, char *end, int proto) {
+	if (!strncmp(end, "/udp", 4)) {
+		if (proto == IPPROTO_TCP && proto != -1)
+			return 0;
+		buf.protocol = IPPROTO_UDP;
+		buf.socktype = SOCK_DGRAM;
+	} else if (!strncmp(end, "/tcp", 4)) {
+		if (proto == IPPROTO_UDP && proto != -1)
+			return 0;
+		buf.protocol = IPPROTO_TCP;
+		buf.socktype = SOCK_STREAM;
+	} else {
+		return 0;
+	}
+
+	//TODO(geert): also parse aliases.
+
+	return 1;
+}
+
+static int lookup_serv_file_port(service_result &buf, int proto, int port) {
+	auto file = fopen(_PATH_SERVICES, "r");
+	if (!file) {
+		switch (errno) {
+			case ENOENT:
+			case ENOTDIR:
+			case EACCES:
+				return -EAI_SERVICE;
+			default:
+				return -EAI_SYSTEM;
+		}
+	}
+
+	char line[128];
+	int name_length = 0;
+	while(fgets(line, 128, file)) {
+		char *pos;
+		// easy way to handle comments, just move the end of the line
+		// to the beginning of the comment
+		if ((pos = strchr(line, '#'))) {
+			*pos++ = '\n';
+			*pos = '\0';
+		}
+
+		char *end = NULL;
+		for (pos = line; *pos; pos++) {
+			for (; isalpha(*pos); pos++);
+			int rport = strtoul(pos, &end, 10);
+			if (rport != port || rport > 65535) {
+				pos = end;
+				continue;
+			}
+
+			// We have found the port, time to rewind to the start
+			// of the line.
+			for (; pos[-1]; pos--)
+				name_length++;
```

Comment:
Is the character before the line buffer guaranteed to be 0?

Reply:
Good point, it actually isn't, but I can fix that real quick. Is the rest okay?

Reply:
The `name_length` calculation seems like it includes the whitespace between the name and port as well, is that expected?

Reply:
Hm that's true, I guess that isn't expected. I'll add a check for whitespace too.